testdir.makepyfile fails to write contents with `utf-8` on Python 2.7

Versions:

    Python 2.7.13
    pytest==3.2.2 (master)
    Linux L251 4.4.0-92-generic #115-Ubuntu x86_64 x86_64 x86_64 GNU/Linux

The class `_pytest.pytester.Testdir._makefile` crashes with `UnicodeDecodeError` when trying to write Python with `utf-8` string.

There's a test that reproduces the bug:

``` python
    @pytest.mark.xfail("sys.version_info[0] < 3")
    def test_setup_fixture_error_with_mixed_encoding(self, testdir):
        testdir.makepyfile("""
            # coding: utf-8
            def mixed_encoding(as_unicode, as_utf8):
                assert 0
            def setup_function(function):
                mixed_encoding(u"'São Paulo'", "'S\xc3\xa3o Paulo'")
            def test_nada():
                pass
        """)
        result = testdir.runpytest()
        result.stdout.fnmatch_lines([
            "*ERROR at setup of test_nada*",
            "*setup_function(function):*",
            "*mixed_encoding(as_unicode, as_utf8):*",
            "*assert 0*",
            "*1 error*",
        ])
        assert result.ret != 0
        if sys.version_info[0] >= 3:
            result.stdout.fnmatch_lines([
                "*as_unicode = \"'São Paulo'\", as_utf8 = \"'SÃ£o Paulo'\"*",
            ])
        else:
            result.stdout.fnmatch_lines([
                "*as_unicode = \"'São Paulo'\", as_utf8 = \"'São Paulo'\"*",
            ])
```

Resulting in:

```
/opt/loggi/pytest/.tox/py27/lib/python2.7/site-packages/_pytest/pytester.py:550: in makepyfile
    return self._makefile('.py', args, kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Testdir local('/tmp/pytest-of-fernando/pytest-4/testdir/test_setup_fixture_error_with_mixed_encoding0')>, ext = '.py'
args = ('\n            # coding: utf-8\n            def mixed_encoding(as_unicode, as_utf8):\n                assert 0\n     ...oding(u"\'S\xc3\xa3o Paulo\'", "\'S\xc3\xa3o Paulo\'")\n            def test_nada():\n                pass\n        ',)
kwargs = {}, encoding = 'utf-8'

    def _makefile(self, ext, args, kwargs, encoding="utf-8"):
        items = list(kwargs.items())
        if args:
            source = py.builtin._totext("\n").join(
>               map(py.builtin._totext, args)) + py.builtin._totext("\n")
E           UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 184: ordinal not in range(128)

/opt/loggi/pytest/.tox/py27/lib/python2.7/site-packages/_pytest/pytester.py:493: UnicodeDecodeError
```

--- a/_pytest/pytester.py
+++ b/_pytest/pytester.py
@@ -480,29 +480,24 @@ class Testdir:
         if not hasattr(self, '_olddir'):
             self._olddir = old
 
-    def _makefile(self, ext, args, kwargs, encoding="utf-8"):
+    def _makefile(self, ext, args, kwargs, encoding='utf-8'):
         items = list(kwargs.items())
+
+        def to_text(s):
+            return s.decode(encoding) if isinstance(s, bytes) else six.text_type(s)
+
         if args:
-            source = six.text_type("\n").join(
-                map(six.text_type, args)) + six.text_type("\n")
+            source = u"\n".join(to_text(x) for x in args)
             basename = self.request.function.__name__
             items.insert(0, (basename, source))
+
         ret = None
-        for name, value in items:
-            p = self.tmpdir.join(name).new(ext=ext)
+        for basename, value in items:
+            p = self.tmpdir.join(basename).new(ext=ext)
             p.dirpath().ensure_dir()
             source = Source(value)
-
-            def my_totext(s, encoding="utf-8"):
-                if isinstance(s, six.binary_type):
-                    s = six.text_type(s, encoding=encoding)
-                return s
-
-            source_unicode = "\n".join([my_totext(line) for line in source.lines])
-            source = six.text_type(source_unicode)
-            content = source.strip().encode(encoding)  # + "\n"
-            # content = content.rstrip() + "\n"
-            p.write(content, "wb")
+            source = u"\n".join(to_text(line) for line in source.lines)
+            p.write(source.strip().encode(encoding), "wb")
             if ret is None:
                 ret = p
         return ret

--- a/changelog/2738.bugfix
+++ b/changelog/2738.bugfix
@@ -1,0 +1,1 @@
+Internal ``pytester`` plugin properly encodes ``bytes`` arguments to ``utf-8``.

--- a/testing/test_pytester.py
+++ b/testing/test_pytester.py
@@ -1,3 +1,4 @@
+# -*- coding: utf-8 -*-
 from __future__ import absolute_import, division, print_function
 import pytest
 import os
@@ -118,6 +119,16 @@ def test_makepyfile_unicode(testdir):
     except NameError:
         unichr = chr
     testdir.makepyfile(unichr(0xfffd))
+
+
+def test_makepyfile_utf8(testdir):
+    """Ensure makepyfile accepts utf-8 bytes as input (#2738)"""
+    utf8_contents = u"""
+        def setup_function(function):
+            mixed_encoding = u'São Paulo'
+    """.encode('utf-8')
+    p = testdir.makepyfile(utf8_contents)
+    assert u"mixed_encoding = u'São Paulo'".encode('utf-8') in p.read('rb')
 
 
 def test_inline_run_clean_modules(testdir):

